url-plugins/src/main/java/co/cask/hydrator/plugin/realtime/UrlFetchRealtimeSource.java

```diff
@@ -0,0 +1,130 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.realtime;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.realtime.RealtimeSource;
+import co.cask.cdap.etl.api.realtime.SourceState;
+import co.cask.hydrator.plugin.realtime.config.UrlFetchRealtimeSourceConfig;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.io.CharStreams;
+
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Real Time Source to fetch data from a url.
+ */
+@Plugin(type = RealtimeSource.PLUGIN_TYPE)
+@Name("URLFetch")
+@Description("Fetch data from an external URL at a regular interval.")
+public class UrlFetchRealtimeSource extends RealtimeSource<StructuredRecord> {
+  private static final int TIMEOUT = 60 * 1000;
+  private static final String METHOD = "GET";
+  private static final String POLL_TIME_STATE_KEY = "lastPollTime";
+  private static final Schema SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("url", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("responseCode", Schema.of(Schema.Type.INT)),
+    Schema.Field.of("headers", Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING))),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  private final UrlFetchRealtimeSourceConfig config;
+
+  public UrlFetchRealtimeSource(UrlFetchRealtimeSourceConfig config) {
+    this.config = config;
+  }
+
+  @Nullable
+  @Override
+  public SourceState poll(Emitter<StructuredRecord> writer, SourceState currentState) throws Exception {
+    byte[] lastPollTimeBytes = currentState.getState(POLL_TIME_STATE_KEY);
+    if (lastPollTimeBytes != null) {
+      long lastPollTime = Bytes.toLong(lastPollTimeBytes);
+      long currentPollTime = System.currentTimeMillis() / 1000;
+      long diff = currentPollTime - lastPollTime;
+      if (config.getIntervalInSeconds() - diff > 0) {
+        TimeUnit.SECONDS.sleep(1L);
```

Comment:
you could do this a little more fine-grained by calculating the diff in millis instead of in seconds, then sleeping for the amount required to get to the interval